st2common/st2common/services/trace.py

```diff
@@ -0,0 +1,258 @@
+from st2common.models.db.trace import TraceDB, TraceComponentDB
+from st2common.persistence.execution import ActionExecution
+from st2common.persistence.trace import Trace
+
+LOG = logging.getLogger(__name__)
+
+__all__ = [
+    'get_trace_db_by_action_execution',
+    'get_trace_db_by_rule',
+    'get_trace_db_by_trigger_instance',
+    'get_trace',
+    'add_or_update_given_trace_context',
+    'add_or_update_given_trace_db'
+]
+
+
+def _get_valid_trace_context(trace_context):
+    """
+    Check if tarce_context is a valid type and returns a TraceContext object.
+    """
+    assert isinstance(trace_context, (TraceContext, dict))
+
+    # Pretty much abuse the dynamic nature of python to make it possible to support
+    # both dict and TraceContext types.
+    if isinstance(trace_context, dict):
+        trace_context = TraceContext(**trace_context)
+
+    return trace_context
+
+
+def _get_single_trace_by_component(**component_filter):
+    """
+    Tries to return a single Trace mathing component_filter. Raises an exception
+    when a filter matches multiple.
+    """
+    traces = Trace.query(**component_filter)
+    if len(traces) == 0:
+        return None
+    elif len(traces) > 1:
+        raise UniqueTraceNotFoundException(
+            'More than 1 trace matching %s found.' % component_filter)
+    return traces[0]
+
+
+def get_trace_db_by_action_execution(action_execution=None, action_execution_id=None):
+    if action_execution:
+        action_execution_id = str(action_execution.id)
+    return _get_single_trace_by_component(action_executions__object_id=action_execution_id)
+
+
+def get_trace_db_by_rule(rule=None, rule_id=None):
+    if rule:
+        rule_id = str(rule.id)
+    # by rule could return multiple traces
+    return Trace.query(rules__object_id=rule_id)
+
+
+def get_trace_db_by_trigger_instance(trigger_instance=None, trigger_instance_id=None):
+    if trigger_instance:
+        trigger_instance_id = str(trigger_instance.id)
+    return _get_single_trace_by_component(trigger_instances__object_id=trigger_instance_id)
+
+
+def get_trace(trace_context, ignore_trace_id=False):
+    """
+    :param trace_context: context object using which a trace can be found.
+    :type trace_context: ``dict`` or ``TraceContext``
+
+    :param ignore_trace_id: Even if a trace_id is provided will be ignored.
+    :type ignore_trace_id: ``str``
+
+    :rtype: ``TraceDB``
+    """
+
+    trace_context = _get_valid_trace_context(trace_context)
+
+    if not trace_context.id_ and not trace_context.trace_id:
+        raise ValueError('Atleast one of id_ or trace_id should be specified.')
+
+    if trace_context.id_:
+        try:
+            return Trace.get_by_id(trace_context.id_)
+        except (ValidationError, ValueError):
+            LOG.warning('Database lookup for Trace with id="%s" failed.',
+                        trace_context.id_, exc_info=True)
+            raise StackStormDBObjectNotFoundError(
```

Comment:
Not really a blocker for this PR, but shouldn't `StackStormDBObjectNotFoundError` really be thrown in the lower layer? (IIRC, we also talked about making some of those changes, right?)